weather_dl/download_pipeline/partition.py

```diff
@@ -125,19 +133,45 @@ def skip_partition(config: Config, store: Store) -> bool:
     return False
 
 
-def prepare_partitions(config: Config) -> t.Iterator[Config]:
-    """Iterate over client parameters, partitioning over `partition_keys`.
+def prepare_partition_index(config: Config,
+                            n_chunks: t.Optional[int] = None) -> t.Iterator[t.Tuple[Config, t.List[t.Tuple[int]]]]:
+    """Produce indexes over client parameters, partitioning over `partition_keys`
 
     This produces a Cartesian-Cross over the range of keys.
 
     For example, if the keys were 'year' and 'month', it would produce
     an iterable like:
+        ( (0, 0), (0, 1), (0, 2), ...)
+
+    After the indexes were converted back to keys, it would produce values like:
         ( ('2020', '01'), ('2020', '02'), ('2020', '03'), ...)
 
     Returns:
-        An iterator of `Config`s.
+        An iterator of index tuples.
     """
-    for option in itertools.product(*[config.selection[key] for key in config.partition_keys]):
+    dims = [range(len(config.selection[key])) for key in config.partition_keys]
+    n_partitions = math.prod([len(d) for d in dims])
+    logger.info(f'Creating {n_partitions} partitions.')
+
+    if n_chunks is None:
+        n_chunks = 50
+
+    chunk_size = n_partitions // n_chunks + 1
```

Comment:
From a usability perspective, I would consider reversing this logic: figure out the number of partitions from the desired chunk sizes. This would probably require a small refactor, but I'm pretty sure you always know how exactly how many configs are to be processed when creating a job?

For example, if each worker processes ~10 configs/second, and you don't want this stage to need to take more than a small number of minutes, you might pick a chunk size in the range of 1000 to 10000. Smaller is probably fine/better (given Beam's autoscaling), as long as overhead is minimized.

Reply:
I appreciate this feedback -- I was on the fence about which way this should go. 

As far as smaller is better: at a certain point, too small of chunks will be slower due to the reshuffle overhead. 1000 to 10_000 chunks seem to work well in practice.